File: interviewbit-trees-max-depth-of-binary-tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque


class Solution:
    # @param A : root node of tree
    # @return an integer
    def maxDepth(self, A):
        if not A:
            return 0
        max_d = 1
        depth = 1
        queue = deque([[A, depth]])
        while queue:
            current, depth = queue.popleft()
            max_d = max(max_d, depth)
            if current.left:
                queue.append([current.left, depth + 1])
            if current.right:
                queue.append([current.right, depth + 1])

        return max_d

File: interviewbit-trees-max-depth-of-binary-tree.java
/**
 * Definition for binary tree
 * class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int maxDepth(TreeNode a) {
        if (a == null) {
            return 0;
        } else if (a.left == null && a.right == null) {
            return 1;
        } else {
            return Math.max(maxDepth(a.left) + 1, maxDepth(a.right) + 1);
        }
    }
}
